rpk: brokers list exposing Host/Port/Rack/UUID

daisukebe opened the pull request:
Relevant to https://github.com/redpanda-data/common-go/pull/27.

`rpk redpanda admin brokers list` now exposes additional fields that makes troubleshooting easier.

Before:
```
NODE-ID  NUM-CORES  MEMBERSHIP-STATUS  IS-ALIVE  BROKER-VERSION
0        2          active             true      v24.2.3 - be49068f63acc31d34b9d186cdfa62087fd99222
1        2          active             true      v24.2.3 - be49068f63acc31d34b9d186cdfa62087fd99222
2        2          active             true      v24.2.3 - be49068f63acc31d34b9d186cdfa62087fd99222
```

After:
```
ID    HOST       PORT   RACK  CORES  MEMBERSHIP  IS-ALIVE  VERSION  UUID
0     localhost  33146  a     2      active      true      24.2.3   e4ff0f11-6fc2-4df1-b129-bbf2630e7c18
1     localhost  33147  -     2      active      true      24.2.3   8b4649c4-7a92-4818-a93e-a3a244fad6e5
3     localhost  33148  -     2      active      true      24.2.3   b00efb2e-6beb-46b1-86f8-6fa2566bfa9b
```
It uses the fairly new `/v1/broker_uuids` endpoint for the UUID field. If a cluster doesn't support the endpoint, the command just hide the column. 

If multiple node UUIDs being found for a node ID, the command prints all with a comma separated fashion in the same row. It could happen when a new node joins the cluster using the same node ID, e.g. being hard-corded in redpanda.yaml. This is not encouraged though.

```
ID    HOST       PORT   RACK  CORES  MEMBERSHIP  IS-ALIVE  VERSION  UUID
0     localhost  33146  a     2      active      true      24.2.3   e4ff0f11-6fc2-4df1-b129-bbf2630e7c18
1     localhost  33147  -     2      active      true      24.2.3   8b4649c4-7a92-4818-a93e-a3a244fad6e5
3     localhost  33148  -     2      active      true      24.2.3   b00efb2e-6beb-46b1-86f8-6fa2566bfa9b, 88fc966d-1b6c-470b-beb7-fa3ba162f998
```

With the new `--include-decommissioned` flag, it can list all decommissioned brokers along with its ID and UUID. This is useful when we deal with the `node_id_overrides` feature, https://github.com/redpanda-data/redpanda/pull/22972

```
ID    HOST       PORT   RACK  CORES  MEMBERSHIP  IS-ALIVE  VERSION  UUID
0     localhost  33146  a     2      active      true      24.2.3   e4ff0f11-6fc2-4df1-b129-bbf2630e7c18
1     localhost  33147  -     2      active      true      24.2.3   8b4649c4-7a92-4818-a93e-a3a244fad6e5
3     localhost  33148  -     2      active      true      24.2.3   b00efb2e-6beb-46b1-86f8-6fa2566bfa9b, 88fc966d-1b6c-470b-beb7-fa3ba162f998
2     -          -      -     -      -           -         -        d671a64b-02c2-4bbf-8437-73deac4db43a
```

Added a detailed help text too.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.2.x
- [x] v24.1.x
- [ ] v23.3.x

## Release Notes

### Improvements

* rpk: `redpanda admin brokers list` exposes Host/Port/Rack/UUID additionally

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/55629#01924c7e-d38f-428e-9563-5089462bc888
<!--ba967369-9782-4655-91df-ab99c54a8e5f-->

Deflaimun commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -14,7 +18,27 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
-		Args:    cobra.ExactArgs(0),
+		Long: `List the brokers in your cluster.
+
+This command lists all brokers in the cluster whatever a broker is up or down,
```
Not sure if up/down can be changed to active/inactive. 
```suggestion
This command lists all brokers in the cluster, regardless of whether they are active or inactive, unless they have been decommissioned.
```

Deflaimun commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -14,7 +18,27 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+This command lists all brokers in the cluster whatever a broker is up or down,
+unless it is decommissioned.
+
+The output table carries the following columns:
```
```suggestion
The output table contains the following columns:
```

Deflaimun commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -14,7 +18,27 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+MEMBERSHIP       Whether a broker is decommissioned or not
+IS-ALIVE         Whether a broker is alive or offline
+VERSION          Broker version
+UUID (Optional)  Yet another exclusive identifier for a broker
```
```suggestion
UUID (Optional)  Additional broker exclusive identifier
```

Deflaimun commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -14,7 +18,27 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+The output table carries the following columns:
+
+ID               Node ID, an exclusive identifier for a broker
+HOST             Internal RPC address
```
Does users recognize what is a RPC address? Do we have another name that we can use here ? Like "host internal API address" or something.

daisukebe replied:
`RPC address` itself is common but I should have made it clearer on what it actually is. How about `Internal RPC address for communication between nodes`?

Deflaimun replied:
`Internal RPC address for communication between brokers`

daisukebe replied:
Thank you.

micheleRP commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -14,7 +18,27 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+VERSION          Broker version
+UUID (Optional)  Yet another exclusive identifier for a broker
+
+NOTE: The UUID column will be hidden when the cluster doesn't expose the UUID
+info via admin API or the API call fails to rertieve UUIDs.
```
```suggestion
NOTE: The UUID column is hidden when the cluster doesn't expose the UUID in the Admin API, or the API call fails to retrieve UUIDs.
```

twmb commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -26,22 +50,25 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
 
 			args := func(b *rpadmin.Broker) []interface{} {
-				ret := []interface{}{b.NodeID, b.NumCores, b.MembershipStatus}
+				ret := []interface{}{b.NodeID, b.InternalRPCAddress, b.InternalRPCPort, b.Rack, b.NumCores, b.MembershipStatus, *b.IsAlive, extractVersion(b.Version)}
```
If rack is empty, I suggest replacing with `-`, so that the output is consistent for awk.

daisukebe commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -14,7 +18,27 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+ID               Node ID, an exclusive identifier for a broker
+HOST             Internal RPC address
+PORT             Internal RPC port
+RACK             Assigned rack ID
```
As Travis suggested I will print `-` when rack is not configured. Thanks!

oleiman commented:
/ci-repeat 1
skip-units
skip-redpanda-build

oleiman commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -33,7 +33,7 @@ CORES            Number of cores (shards) on a broker
 IS-ALIVE         Whether a broker is alive or offline
 VERSION          Broker version
-UUID (Optional)  Yet another exclusive identifier for a broker
+UUID (Optional)  Additional broker exclusive identifier
```
nit: phrasing (I think the original was more precise, modulo "yet")

```suggestion
UUID (Optional)  Additional exclusive identifier for a broker
```

daisukebe commented:
Rebased. In https://github.com/redpanda-data/redpanda/commit/2d1d078a15b66229a190daeb5e1cdd2be64058b8, I have made the following changes. 

* Accepted help text suggestions
* Returning `-` when rack is not configured
* Added the `--all` flag to show decommissioned node ID & UUID, which is helpful when we deal with the `node_id_overrides` feature, https://github.com/redpanda-data/redpanda/pull/22972

```
ID    HOST       PORT   RACK  CORES  MEMBERSHIP  IS-ALIVE  VERSION  UUID
0     localhost  33146  a     2      active      true      v24.2.3  e4ff0f11-6fc2-4df1-b129-bbf2630e7c18
1     localhost  33147  -     2      active      true      v24.2.3  8b4649c4-7a92-4818-a93e-a3a244fad6e5
3     localhost  33148  -     2      active      true      v24.2.3  b00efb2e-6beb-46b1-86f8-6fa2566bfa9b
2     -          -      -     -      -           -         -        d671a64b-02c2-4bbf-8437-73deac4db43a
```

daisukebe commented:
Nit change; I should have considered the case where `/v1/broker_uuids` isn't supposed when `--all` is used.

daisukebe commented:
/ci-repeat 1
skip-units
skip-redpanda-build

r-vasquez commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -26,27 +50,88 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+
+// extractVersion gets rid of a commit hash from a version string.
+// e.g. Given "v24.2.3 - be49068f63acc31d34b9d186cdfa62087fd99222", "v24.2.3" returns.
+func extractVersion(fullVersion string) string {
+	re := regexp.MustCompile(`v\d+\.\d+\.\d+`)
+	return re.FindString(fullVersion)
+}
```
Can you try using the `version` package? https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/redpanda/version.go

r-vasquez commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -26,27 +50,88 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+		}
+	}
+	if len(UUIDs) == 0 {
+		return "UUID Not Found"
```
This means we will have something like

```
ID    HOST       PORT   RACK  CORES  MEMBERSHIP  IS-ALIVE  VERSION  UUID
0     localhost  33146        2      active      true      v24.2.3  UUID Not Found
```

Wonder if we should use `-` as in RACK, or decommissioned nodes. In which cases will we have empty node UUIDs?

daisukebe replied:
Rack is optional whereas missing UUID is an unexpected scenario, which indicates a core issue. Thus I wanted to distinguish each other. Should I make it more explicit when UUID is not found?

daisukebe replied:
It should be something like `FATAL: UUID Not Found`?

daisukebe replied:
After some thought, I’m leaning towards using a hyphen when a UUID is not found to make it more awk-friendly. I believe it is enough to recognize the issue with hyphen.

r-vasquez replied:
Great, thanks!

r-vasquez reviewed (commented):
LGTM, 2 minor comments/questions.

Thanks for this and for the detailed description of the command and the changes

daisukebe commented:
In https://github.com/redpanda-data/redpanda/commit/bf93c5dd3324c59554ed94d51387ba9050f9fb92, I've made the following changes:

* Use `redpanda.Version` to extract the broker version
* Return hyphen when UUID is not found
* Change the flag from `--all` to `--include-decommissioned / -d` to make it more explicit

daisukebe commented:
Sorry, did another rebase; merged `checkEmpty` and `formatVersion` to one func `formatOutput`.

r-vasquez commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -1,20 +1,44 @@
+		Long: `List the brokers in your cluster.
+
+This command lists all brokers in the cluster, active and inactive, unless they have been decommissioned.
+Using the "--all" flag, it lists decommissioned brokers with associated UUIDs too.
```
This changed in your latest commit
```suggestion
Using the "--include-decommissioned" flag, it lists decommissioned brokers with associated UUIDs too.
```

r-vasquez commented on src/go/rpk/pkg/cli/redpanda/admin/brokers/list.go:
```diff
@@ -26,27 +50,82 @@ func newListCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+			}
 		},
 	}
+	cmd.Flags().BoolVarP(&decom, "include-decommissioned", "d", false, "If true, include decommissioned brokers")
```
Nice! I like this change :+1:

r-vasquez reviewed (commented):
LGTM, just one small left over in the docs from the flag name change. Thanks!

r-vasquez reviewed (approved):
Thanks!

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-23595-v24.1.x-693 remotes/upstream/v24.1.x
git cherry-pick -x 08e30aa37d
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/11245335948).
